README.md
# Moshier Ephemeris ES6 Re-implementation

This is a "re-implementation" of version 0.1.0 of Mivion's Moshier Ephemeris javascript implementation (found here: https://github.com/mivion/ephemeris).

The goal is to re-implement the codebase with ES6 modules, classes, and refactored programming to promote better debugging / testing, readability, and extensibility of the code.

This amazing work done by Moshier and Mivion deserves a lot of love. Hope you enjoy it!

##  Description

ES6 re-implementation of ephemeris calculations for sun, planets, comets, asteroids and stars.

This implementation based on Steve Moshier (http://www.moshier.net).

Licensed under GPL version 3 (https://www.gnu.org/licenses/gpl-3.0.html).

## About Moshier's Ephemeris

An ephemeris is a table that gives you the position of various celestial bodies in the solar system for every minute of every day of every year. Before computers, they were frequently sold in gigantic books every decade or so.

Moshier's ephemeris is a set of formula written by Stephen Moshier that generates the apparent positions of celestial bodies when given a particular date/time/location, without all the tables.

The Jet Propulsion lab also has an ephemeris, which is only accessible via their website or a telnet connection. There's also the "Swiss Ephemeris", which is available for free until you hit a certain threshold.

Both of these popular ephemerides aren't as distributable as Moshier's. They're both essentially gigantic tables - dozens of MBs large - or otherwise hard to readily access (see: telnet connection). Moshier's, with its tiny size, is built for the modern web and can be implemented directly into something as small as a phone, raspberry pi, or any client side app.

Moshier's Ephemeris is good from -3000 B.C.E - 3000 C.E. Its results are always within less than a degree of the other leading Ephemerii, making this a highly precise library.

## Celestial Bodies included
- Sun
- Moon
- Earth
- Mercury
- Venus
- Mars
- Jupiter
- Saturn
- Uranus
- Neptune
- Pluto
- Chiron
- Sirius

## Extra features
- Thoroughly tested code & an easy to implement `Ephemeris` class for external use.
- Moon phase descriptions
- Lunar nodes & perigee / apogee calculations
- Planetary & asteroid retrograde motion reporting, as well as optional shadow phase determinations.

## Future work
- Eclipses
- Table generator for retrograde / eclipse events within a user-specified timeframe.
- Add more comets & asteroids


## Performance
- The ephemeris is extremely fast and performs all of its calculations in under what appears to be 100ms. I haven't actually benchmarked things but hope to in the future.
- As of 1.2.0, I introduced retrograde shadow calculations which introduce a significant performance impact (nearly 10x slowdown). I made these calculations optional and set them to be turned off by default. Calculating retrograde shadows require the ephemeris to calculate the next and previous retrograde & direct dates for each body, which I'm currently doing by brute force since I don't know if any formula exist for these sorts of calculations. If anyone knows, please reach out!

## Demo

Open the file: `/demo/index.html` in your browser.

## Usage

##### Load the library
```
// With ES6 imports

import Ephemeris from './build/ephemeris-1.2.0.bundle.js'

const ephemeris = new Ephemeris({...date/location})

ephemeris.Results
```

```
// or in a browser
<script type='text/javascript' src='./ephemeris-1.2.0.bundle.js' charset='utf-8'></script>

<script>
  const ephemeris = new Ephemeris.default({...date/location})

  ephemeris.Results
</script>
```

#####  Create a new ephemeris instance for all celestial bodies
```
# Create a new ephemeris instance

// January 1st, 2000, 0:00 UTC - Cambridge, MA

// NOTE - months go from 0 - 11 (0 = jan)

const ephemeris = new Ephemeris({
  year: 2000, month: 0, day: 1, hours: 0, minutes: 0, latitude: 41.37, longitude: -71.1, calculateShadows: false
})
```

##### Generate ephemeris for a single body ( or multiple )
```
# a single body = with key: "string"

const ephemeris = new Ephemeris({
  key: "jupiter",
  year: 2000, month: 0, day: 1, hours: 0, minutes: 0,
  latitude: 41.37, longitude: -71.1,
  calculateShadows: false
})

# multiple specific bodies with key: [array]

const ephemeris = new Ephemeris({
  key: ["jupiter", "venus", "moon", "chiron"],
  year: 2000, month: 0, day: 1, hours: 0, minutes: 0,
  latitude: 41.37, longitude: -71.1,
  calculateShadows: false
})
```

##### View all results
```
# View all results

ephemeris.Results

// => Array[{sun}, {moon}, {mercury}...]
```

##### Get the Body Position and Retrograde motion

```
# Get position

ephemeris.mercury.position

# => {
  position: {
    ...
    apparentLongitude: 274.38206441358966,
    apparentLongitude30String: "4°22'55"",
    apparentLongitudeString: "274°22'55"",
    ...
  }
}

# Get motion

ephemeris.mercury.motion

# => {
  isRetrograde: false,
  oneSecondMotionAmount: 0.00000031751...,
  withinPreRetrogradeShadow: false, // optional
  withinPostRetrogradeShadow: false // optional
}
```

##### complete body example

```
# View a specific celestial body result (if generated)

ephemeris.mercury

// => {
  aberration: {
    dRA: -1.9269831284660512, dDec: 1.7324642445063785
  },
  anomaly: 198.7199,
  dailyMotion : 4.09236,
  distance : 0.4662701071857169,
  eccentricity: 0.205628,
  epoch: 2458849.491717961,
  equinox: {
    julian: 2451545
  },
  equinoxEclipticLonLat: {
    0: 4.580688286536208, 1: -0.06840241722458161, 2: 0.46626601943095985, 3: {…}, 4: {…}
  },
  inclination: 7.0048,
  key: "mercury",
  lightTime: 11.926136235901744,
  longitude: 4.575414616860342,
  magnitude: -0.42,
  node: 48.177,
  perihelion: 29.074,
  position: {
    aberration: {
      dRA: -1.520496036316089, dDec: -0.6041546957957655
    }
    altaz: {
      atmosphericRefraction: {
        deg: 0, dRA: 0, dDec: -6.869998111961196e-11
      }
      dLocalApparentSiderialTime: 0.5064527239416776,
      diurnalAberation: {
        ...
      },
      diurnalParallax: {
        ...
      },
      localApparentSiderialTime: {...},
      topocentric: {
        ...
      },
      transit: {
        ...
      }
    },
    apparent: {
      ...
    },
    apparentGeocentric: {
    ...
    },
    apparentLongitude: 274.38206441358966,
    apparentLongitude30String: "4°22'55"",
    apparentLongitudeString: "274°22'55"",
    approxVisual: {
      magnitude: -1.281820354968265, phase: 0.9886399799797707
    },
    astrometricB1950: {
      ...
    },
    astrometricJ2000: {
      ...
    },
    constellation: "Sgr Sagittarii",
    date: {
      year: 2020, month: 1, day: 1, hours: 0, minutes: 0
    },
    deflection: {
      sunElongation: 5.768325480048136, lightDeflection: {…}
    },
    equatorialDiameter: 4.686223488825094,
    geocentricDistance: 1.4339905077136652,
    nutation: {
      dRA: -1.2125665235304237, dDec: 1.1499169000116969
    }
    polar: [4.575812495945637, -0.06835690545143856, 0.4662660194309599],
    rect: [-0.06351901141088291, -0.4101381667492277, -0.21250842858566044],
    trueGeocentricDistance: 1.4340239958103242
  },
  motion: {
    isRetrograde: false,
    oneSecondMotionAmount: 0.00000031751...,
    withinPreRetrogradeShadow: false, // optional
    withinPostRetrogradeShadow: false // optional
  }
}
```

##### Get the Earth and Observer
```
# Get the Earth and Observer

ephemeris.Earth

// => Object{earth}

ephemeris.Observer

// => Object{observer}
```

##### Get Moon Phase / Nodes / Orbit readings

```
// Phases
ephemeris.moon.position

// => {
  ...
  illuminatedFraction: 0.3992198452219991, // 0...1 -- 0 = none, 0.5 = half, 1 = full
  phaseDaysBefore: 1.000054577340336,
  phasedaysPast: undefined,
  phaseDaysDistance: 1.000054577340336,
  phaseDecimal: 0.2171554039712993, // 0...1 -- 0 = new, 0.25 = first q, 0.5 = full, 0.75 = last q
  phaseQuarter: 0,
  phaseQuarterString: "New Moon",
  quarterApproximationDirectionString: "Entering",
  quarterApproximationString: "First Quarter",
  shapeDirectionString: "Waxing",
  shapeString: "Crescent"
  ...
}

// => Object{luna}

ephemeris.moon.orbit

// => {
  meanApogee: { // aka lilith
    apparentLongitude: 357.0998704627402,
    apparentLongitudeString: "357°6'0"",
    apparentLongitude30String: "27°6'0""
  },
  meanAscendingNode: {...}, // aka north node
  meanDescendingNode: {...}, // aka south node
  meanPerigee: {...}
}
```

## Installation

```
npm install

// or yarn install
```

## Testing

```
npm test

// or yarn test
```

### Notes on precision & accuracy upgrading from 0.1.0 to 1.0.0 and beyond

1) There is a very tiny differences (in the magnitude of 0.0000005 degrees) between the `apparentLongitude` decimal calculation of the `Sun` in the 1.0.0 implementation vs the 0.1.0 implentation. I've tracked this down to a specific calculation - `epsilon.js`.

I believe this upgrade actually fixed a bug in the original implementation. The bug was centered around the way `epsilon.js` was implemented. In the 0.1.0 implementation, `$moshier.epsilon` was a global variable that was reassigned frequently. This, I believe, resulted in unintentional mutations.

I refactored the code to treat this as an immutable locally scoped object, because I do not believe `epsilon` was intended to store global mutations of its data.

Legacy code (mutations noted):

`v0.1.0 Sun.js - line 76 - 87`
```
$moshier.epsilon.calc ($moshier.body.earth.position.date); // Sets $moshier.epsilon - now locally scoped & immutable in 1.0.0
$moshier.nutation.calc ($moshier.body.earth.position.date, ecr); // mutates $moshier.epsilon

/* Display the final apparent R.A. and Dec.
 * for equinox of date.
 */
$moshier.body.sun.position.constellation = $moshier.constellation.calc (ecr, $moshier.body.earth.position.date); // mutates $moshier.epsilon

$moshier.body.sun.position.apparent = $util.showrd (ecr, pol);

/* Show it in ecliptic coordinates */
y  =  $moshier.epsilon.coseps * rec[1]  +  $moshier.epsilon.sineps * rec[2]; // utilizes $moshier.epsilon
```

This 0.0000005 difference in the Sun's `apparentLongitude` between versions 0.1.0 and 1.0.0 is probably good.

Everything else appears to be exactly the same according to my tests.

2) Found tiny correction (0.1*e-15 or 0.000000000000001) on moon `illuminatedFraction` and `phaseDaysPast / phaseDaysBefore` as a result of gplan refactoring. This was most likely due to the use of conversion constants for radians-to-seconds and seconds-to-radians (RTS and STR) in the app. One possible solution would be to use equations involving Math.PI instead, however this tiny difference seems unimportant and an issue with javascript's math rather than the refactor.


CHANGELOG.md
# Changelog

## **v1.2.0**

##### Added

- Added `.motion` property to heliocentric bodies for reporting retrograde motion and pre/post shadow phases.
- Added `calculateShadows` key to Ephemeris class constructor to toggle costly shadow calculations on / off.

## **v1.1.0**

##### Fixed

##### Added

- Added `observerInstance.Date.dateString` property which returns the formatted date string of the observer
- Added new moon position properties include phase & shape descriptions and orbit calculations (mean ascending node + mean descending node and mean perigee + mean apogee):

```
new Ephemeris: {
  moon: {
    position: {
      phaseDecimal: float // property which returns a 0.0 - 0.99 representation of the phase.
      phaseDaysDistance: float // positive float indicating distance from closest phase quarter
      shapeDirectionString: string // 'Waxing' or 'Waning' depending on the phaseDecimal value
      shapeString: string // 'Crescent' or 'Gibbous' depending on the illuminatedFraction
      quarterApproximationValue: float // the amount (in days) in either direction of a quarter event to allow for determining the approximate quarter phase. Defaults to 1.5 days in either direction of a quarter event.
      quarterApproximationString: string // QUARTER EVENT<'New Moon', 'First Quarter', 'Full Moon', 'Last Quarter'> determined by the phaseDecimal and quarterApproximationValue
      quarterApproximationDirectionString: string // "Entering" or "Leaving" the quarter approximation
      withinQuarterApproximation: boolean // true/false determined by whether phaseDecimal is inside of quarterApproximationValue
    }
    orbit: {
      meanAscendingNode: {
        apparentLongitude: float
        apparentLongitudeString: string
        apparentLongitude30String: string
      },
      meanDescendingNode: {...},
      meanPerigee: {...},
      meanApogee: {...}
    }
  }
}

```


##### Changed

## **v1.0.0**

##### Fixed
- Fixed Moon `phaseQuarter` bug where integer did not align with expected moon phase start.

##### Added

- Added `phaseQuarterString` property to the `moon` result with a descriptor of the phase ("Full Moon", etc).
- Complete test coverage for all calculated bodies.

##### Changed
- Refactors much of the codebase into ES6 modules, classes, and OOP patterns where easy to do.
- `body.constellation` property returns the name of the constellation instead of the index

##### Breaking Changes

- Added Ephemeris class to access calculations (see `README` for usage examples)
- Requires month range 0 - 11 on instantiation


src/classes/Luna.js
import { DTR, RTD, RTS, REARTH } from '../constants'

import { aberration } from '../utilities/aberration'
import { altaz } from '../utilities/altaz'
import Epsilon from '../utilities/Epsilon'
import { gPlanMoon, get_lp_equinox } from '../utilities/gplan'
import { lonlat } from '../utilities/lonlat'
import { precess } from '../utilities/precess'
import { nutation } from '../utilities/nutation'
import { util } from '../utilities/util'
import { getMeanAscendingNode, getMeanPerigee } from '../utilities/luna'

export default class Luna {
  constructor({body, earthBody, observer, quarterApproximationValue=1.5}={}) {
    this.ra = 0.0 /* Right Ascension */
    this.dec = 0.0 /* Declination */

    this.quarterApproximationValue = quarterApproximationValue

    this._body = this.calculateBody(body, earthBody, observer)

    Object.keys(this._body).forEach(key => {
      this[key] = this._body[key]
    })

    this.orbit = this.calculateOrbit(observer.Date.julian)

    this.calculateBody = this.calculateBody.bind(this)
    this.calcll = this.calcll.bind(this)
    this.calculateOrbit = this.calculateOrbit.bind(this)
    this.calculateMeanAscendingNode = this.calculateMeanAscendingNode.bind(this)
    this.calculateMeanDescendingNode = this.calculateMeanDescendingNode.bind(this)
  }

  static GetPhaseQuarterString(quarterIndex) {
    switch(quarterIndex) {
      case 0:
        return "New Moon"
      case 1:
        return "First Quarter"
      case 2:
        return "Full Moon"
      case 3:
        return "Last Quarter"
      default:
        throw new Error(`Quarter Index: ${quarterIndex} not valid (must be beteen 0 - 3)`)
    }
  }

  static GetShapeString({illuminatedFraction=0.00}={}) {
    if (illuminatedFraction >= 0 && illuminatedFraction < 0.5) return "Crescent"
    if (illuminatedFraction >= 0.5 && illuminatedFraction < 1) return "Gibbous"
  }

  static GetShapeDirectionString({phaseDecimal=0.00}={}) {
    if (phaseDecimal >= 0 && phaseDecimal < 0.5) return "Waxing"
    if (phaseDecimal >= 0.5 && phaseDecimal < 1) return "Waning"
  }

  static GetWithinQuarterApproximation({phaseDaysBefore=0.00, phaseDaysPast=0.00, halfRangeLength=this.quarterApproximationValue}={}) {
    return (!!phaseDaysBefore && phaseDaysBefore <= halfRangeLength) || (!!phaseDaysPast && phaseDaysPast <= halfRangeLength)
  }

  static GetQuarterApproximationString({quarterIndex=0, phaseDaysBefore=0.00, phaseDaysPast=0.00, halfRangeLength=1.5}={}) {
    if (!(!!phaseDaysBefore || !!phaseDaysPast) || (!!phaseDaysBefore && phaseDaysBefore > halfRangeLength) || (!!phaseDaysPast && phaseDaysPast > halfRangeLength)) {
      return
    }

    const closestQuarterIndex = (!!phaseDaysBefore && phaseDaysBefore <= halfRangeLength) ? util.mod(quarterIndex + 1, 4) : util.mod(quarterIndex, 4)

    return `${Luna.GetPhaseQuarterString(closestQuarterIndex)}`
  }

  static GetQuarterApproximationDirectionString({phaseDaysBefore=0.00, phaseDaysPast=0.00, halfRangeLength=1.5}={}) {
    if (!!phaseDaysBefore && phaseDaysBefore <= halfRangeLength) return "Entering"
    if (!!phaseDaysPast && phaseDaysPast <= halfRangeLength) return "Leaving"
  }

  calculateBody(body, earthBody, observer) {
    var i, prtsav; // int
  	var ra0, dec0; // double
  	var x, y, z, lon0; // double
  	var pp = [], qq = [], pe = [], re = [], moonpp = [], moonpol = []; // double

  	body.position = {
  		polar: [],
  		rect: []
  	};

  	/* Geometric equatorial coordinates of the earth.  */
  	for (i = 0; i < 3; i++) {
  		re[i] = earthBody.position.rect [i];
  	}

  	/* Run the orbit calculation twice, at two different times,
  	 * in order to find the rate of change of R.A. and Dec.
  	 */

  	/* Calculate for 0.001 day ago
  	 */


  	this.calcll(observer.Date.julian - 0.001, moonpp, moonpol, body, earthBody, observer); // TDT - 0.001
  	ra0 = this.ra;
  	dec0 = this.dec;
  	lon0 = moonpol[0];
  	/* Calculate for present instant.
  	 */
  	body.position.nutation = this.calcll(observer.Date.julian, moonpp, moonpol, body, earthBody, observer).nutation;

  	body.position.geometric = {
  		longitude: RTD * body.position.polar[0],
  		latitude: RTD * body.position.polar[1],
  		distance: RTD * body.position.polar[2]
  	};

  	/**
  	 * The rates of change.  These are used by altaz () to
  	 * correct the time of rising, transit, and setting.
  	 */
  	body.locals.dradt = this.ra - ra0;
  	if (body.locals.dradt >= Math.PI)
  		body.locals.dradt = body.locals.dradt - 2.0 * Math.PI;
  	if (body.locals.dradt <= -Math.PI)
  		body.locals.dradt = body.locals.dradt + 2.0 * Math.PI;
  	body.locals.dradt = 1000.0 * body.locals.dradt;
  	body.locals.ddecdt = 1000.0*(this.dec-dec0);

  	/* Rate of change in longitude, degrees per day
  	 * used for phase of the moon
  	 */
  	lon0 = 1000.0*RTD*(moonpol[0] - lon0);

  	/* Get apparent coordinates for the earth.  */
  	z = re [0] * re [0] + re [1] * re [1] + re [2] * re [2];
  	z = Math.sqrt(z);
  	for (i = 0; i < 3; i++) {
  		re[i] /= z;
  	}

  	/* aberration of light. */
  	body.position.annualAberration = aberration.calc(re, earthBody, observer, body);

  	/* pe[0] -= STR * (20.496/(RTS*pe[2])); */
  	re = precess.calc(re, observer.Date.julian, -1);
  	nutation.calc(observer.Date, re); // NOTE mutates re

  	for (i = 0; i < 3; i++) {
  		re[i] *= z;
  	}

  	pe = lonlat.calc( re, observer.Date, pe, 0 );

  	/* Find sun-moon-earth angles */
  	for( i=0; i<3; i++ ) {
  		qq[i] = re[i] + moonpp[i];
  	}
  	body = util.angles( moonpp, qq, re, body );

  	/* Display answers
  	 */

  	body.position.apparentGeocentric = {
  		longitude: moonpol [0],
  		dLongitude: RTD * moonpol [0],
  		latitude: moonpol [1],
  		dLatitude: RTD * moonpol [1],
  		distance: moonpol [2] / REARTH
  	};
  	body.position.apparentLongitude = body.position.apparentGeocentric.dLongitude;
  	var dmsLongitude = util.dms(body.position.apparentGeocentric.longitude);
  	body.position.apparentLongitudeString =
  		dmsLongitude.degree + '\u00B0' +
  		dmsLongitude.minutes + '\'' +
  		Math.floor (dmsLongitude.seconds) + '"'
  	;

  	body.position.apparentLongitude30String =
  		util.mod30 (dmsLongitude.degree) + '\u00B0' +
  		dmsLongitude.minutes + '\'' +
  		Math.floor (dmsLongitude.seconds) + '"'
  	;

  	body.position.geocentricDistance = moonpol [2] / REARTH;

  	x = REARTH/moonpol[2];
  	body.position.dHorizontalParallax = Math.asin (x);
  	body.position.horizontalParallax = util.dms (Math.asin (x));

  	x = 0.272453 * x + 0.0799 / RTS; /* AA page L6 */
  	body.position.dSemidiameter = x;
  	body.position.Semidiameter = util.dms (x);

  	x = RTD * Math.acos(-body.locals.ep);
  	/*	x = 180.0 - RTD * arcdot (re, pp); */
  	body.position.sunElongation = x;
  	x = 0.5 * (1.0 + body.locals.pq);
  	body.position.illuminatedFraction = x;

  	/* Find phase of the Moon by comparing Moon's longitude
  	 * with Earth's longitude.
  	 *
  	 * The number of days before or past indicated phase is
  	 * estimated by assuming the true longitudes change linearly
  	 * with time.  These rates are estimated for the date, but
  	 * do not stay body.  The error can exceed 0.15 day in 4 days.
  	 */
  	x = moonpol[0] - pe[0];
  	x = util.modtp ( x ) * RTD;	/* difference in longitude */
  	i = Math.floor (x/90);	/* number of quarters */
    // phaseDecimal - 0.0 - 1.0 float representation of phase.
    body.position.phaseDecimal = ((x / 360) + 0.5) % 1 // 0-0.24 = New, 0.25 - 0.49 = First Quarter, 0.5 - 0.74 = Full, 0.75 - 0.99 = Last Quarter

    x = (x - i*90.0);	/* phase angle mod 90 degrees */

  	/* days per degree of phase angle */
  	z = moonpol[2]/(12.3685 * 0.00257357);

  	if( x > 45.0 ) {
  		y = -(x - 90.0)*z;
  		body.position.phaseDaysBefore = y;
  	} else {
  		y = x*z;
  		body.position.phaseDaysPast = y;
  	}

    body.position.phaseDaysDistance = y

    i = (i+2) % 4;
  	body.position.phaseQuarter = i;
    body.position.phaseQuarterString = Luna.GetPhaseQuarterString(i);

    body.position.shapeString = Luna.GetShapeString({illuminatedFraction: body.position.illuminatedFraction})

    body.position.shapeDirectionString = Luna.GetShapeDirectionString({phaseDecimal: body.position.phaseDecimal})

    body.position.withinQuarterApproximation = Luna.GetWithinQuarterApproximation({
      quarterIndex: body.position.phaseQuarter,
      phaseDaysBefore: body.position.phaseDaysBefore,
      phaseDaysPast: body.position.phaseDaysPast,
      halfRangeLength: this.quarterApproximationValue
    })
    body.position.quarterApproximationString = Luna.GetQuarterApproximationString({
      quarterIndex: body.position.phaseQuarter,
      phaseDaysBefore: body.position.phaseDaysBefore,
      phaseDaysPast: body.position.phaseDaysPast,
      halfRangeLength: this.quarterApproximationValue
    })

    body.position.quarterApproximationDirectionString = Luna.GetQuarterApproximationDirectionString({
      phaseDaysBefore: body.position.phaseDaysBefore,
      phaseDaysPast: body.position.phaseDaysPast,
      halfRangeLength: this.quarterApproximationValue
    })

  	body.position.apparent = {
  		dRA: this.ra,
  		dDec: this.dec,
  		ra: util.hms (this.ra),
  		dec: util.dms (this.dec)
  	};

  	/* Compute and display topocentric position (altaz.c)
  	 */
  	pp[0] = this.ra;
  	pp[1] = this.dec;
  	pp[2] = moonpol[2];
  	body.position.altaz = altaz.calc(pp, observer.Date, body, observer);

    return body
  }

  calcll(julianDate, rect, pol, body, earthBody, observer, result) {
  	var cosB, sinB, cosL, sinL, y, z; // double
  	var qq = [], pp = []; // double
  	var i; // int

  	result = result || {};

  	/* Compute obliquity of the ecliptic, coseps, and sineps.  */
  	const epsilonObject = new Epsilon(julianDate);
  	/* Get geometric coordinates of the Moon.  */
    const lp_equinox = get_lp_equinox(julianDate)
  	rect = gPlanMoon(julianDate, rect, pol, lp_equinox);
  	/* Post the geometric ecliptic longitude and latitude, in radians,
  	 * and the radius in au.
  	 */
  	body.position.polar[0] = pol[0];
  	body.position.polar[1] = pol[1];
  	body.position.polar[2] = pol[2];
  	/* Light time correction to longitude,
  	 * about 0.7".
  	 */
  	pol[0] -= 0.0118 * DTR * REARTH / pol[2];

  	/* convert to equatorial system of date */
  	cosB = Math.cos(pol[1]);
  	sinB = Math.sin(pol[1]);
  	cosL = Math.cos(pol[0]);
  	sinL = Math.sin(pol[0]);
  	rect[0] = cosB*cosL;
  	rect[1] = epsilonObject.coseps*cosB*sinL - epsilonObject.sineps*sinB;
  	rect[2] = epsilonObject.sineps*cosB*sinL + epsilonObject.coseps*sinB;

  	/* Rotate to J2000. */
  	rect = precess.calc( rect, observer.Date.julian, 1 ); // TDT

  	/* Find Euclidean vectors and angles between earth, object, and the sun
  	 */
  	for( i=0; i<3; i++ ) {
  		pp[i] = rect[i] * pol[2];
  		qq[i] = earthBody.position.rect [i] + pp[i];
  	}
  	body = util.angles (pp, qq, earthBody.position.rect, body);

  	/* Make rect a unit vector.  */
  	/* for (i = 0; i < 3; i++) */
  	/*  rect[i] /= EO; */

  	/* Correct position for light deflection.
  	 (Ignore.)  */
  	/* relativity( rect, qq, rearth ); */

  	/* Aberration of light.
  	 The Astronomical Almanac (Section D, Daily Polynomial Coefficients)
  	 seems to omit this, even though the reference ephemeris is inertial.  */
  	/* annuab (rect); */

  	/* Precess to date.  */
  	rect = precess.calc(rect, observer.Date.julian, -1); // TDT

  	/* Correct for nutation at date TDT.
  	 */
    const nutationObject = nutation.getObject({julian: observer.Date.julian})
  	result.nutation = nutation.calc ({julian: observer.Date.julian}, rect); // TDT

  	/* Apparent geocentric right ascension and declination.  */
  	this.ra = util.zatan2(rect[0],rect[1]);
  	this.dec = Math.asin(rect[2]);

  	/* For apparent ecliptic coordinates, rotate from the true
  	 equator into the ecliptic of date.  */

  	cosL = Math.cos(epsilonObject.eps + nutationObject.nuto);
  	sinL  = Math.sin(epsilonObject.eps + nutationObject.nuto);
  	y = cosL * rect[1] + sinL * rect[2];
  	z = -sinL * rect[1] + cosL * rect[2];
  	pol[0] = util.zatan2( rect[0], y );
  	pol[1] = Math.asin(z);

  	/* Restore earth-moon distance.  */
  	for( i=0; i<3; i++ ) {
  		rect[i] *= body.locals.EO;
  	}

  	return result;
  }

  calculateOrbit(julianDate) {
    const orbit = {}

    orbit.meanAscendingNode = util.attachApparentLongitudes({}, getMeanAscendingNode(julianDate))
    orbit.meanDescendingNode = util.attachApparentLongitudes({}, util.mod(orbit.meanAscendingNode.apparentLongitude - 180, 360))

    orbit.meanPerigee = util.attachApparentLongitudes({}, getMeanPerigee(julianDate))
    orbit.meanApogee = util.attachApparentLongitudes({}, util.mod(orbit.meanPerigee.apparentLongitude - 180, 360))

    return orbit
  }

  calculateMeanAscendingNode(julianDate) {
    const meanAscendingNode = {}

    meanAscendingNode.apparentLongitude = getMeanAscendingNode(julianDate)
    meanAscendingNode.apparentLongitudeString = util.decimalDegreesToDMSString(meanAscendingNode.apparentLongitude)
    meanAscendingNode.apparentLongitude30String = util.decimalDegreesToDMSString(util.mod(meanAscendingNode.apparentLongitude, 30))

    return meanAscendingNode
  }

  calculateMeanDescendingNode(ascendingNode) {
    const meanDescendingNode = {}

    meanDescendingNode.apparentLongitude = util.mod(ascendingNode - 180, 360)
    meanDescendingNode.apparentLongitudeString = util.decimalDegreesToDMSString(meanDescendingNode.apparentLongitude)
    meanDescendingNode.apparentLongitude30String = util.decimalDegreesToDMSString(util.mod(meanDescendingNode.apparentLongitude, 30))

    return meanDescendingNode
  }
}
